### PR TITLE
Assignment2 - Conner Hanson

### PR DESCRIPTION
**Pull request by connerhanson**
I think since I removed (rm -r) and re-cloned my repo early on, it shows as if I remade the whole file for each file. I tried to fix this but I'm not exactly sure how I would since I did this at the beginning of the assignment.  Let me know if you have any questions while reviewing. 

After asking for help I think I've mostly fixed the above stated problem.  If it isn't looking right to you, set the Diff View to "Unified" and select "Hide whitespace".

### CODE REVIEW

File: Logger/BaseLoggerMixins.cs [hunk 1]
```diff
@@ -1,6 +1,61 @@
...
+        {
+            throw new System.ArgumentNullException(nameof(logger));
+        }
+        
+        //format string to return, use string.format so the argument is placed into the {0} spot
+        string formattedString = string.Format(CultureInfo.CurrentCulture, message, arguments);
+        
+        //Call log with specified loglevel for method and string
+        logger.Log(LogLevel.Error, formattedString);
+    }
```
Comment (omar5o9) [inline Logger/BaseLoggerMixins.cs:21]:
```suggestion
        if(logger is null)
        {
            throw new System.ArgumentNullException(nameof(logger));
        }
        
        //format string to return, use string.format so the argument is placed into the {0} spot
        string formattedString = string.Format(CultureInfo.CurrentCulture, message, arguments);
        
        //Call log with specified loglevel for method and string
        logger.Log(LogLevel.Error, formattedString);
    }

```
For these methods, instead checking for null and formatting each one, why not try to create a helper method that does all of these things to save you some code by calling the helper in each one instead.

File: Logger/LogFactory.cs [hunk 2]
```diff
@@ -1,10 +1,35 @@
...
+﻿namespace Logger;
+#nullable enable
+
+public class LogFactory
+{
+    //create nullable path string
+    private string? path;   
+
+    //Configure given filePath as path
+    public void ConfigureFileLogger(string filePath)
```
Comment (omar5o9) [inline Logger/LogFactory.cs:10]: This method should be used in the createLogger method
```suggestion
    public void ConfigureFileLogger(string filePath)

```

**Review by omar5o9 — COMMENTED** [review]
There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ❌ (Didn't use use the configure in the CreateLogger method)
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔
All of the above should be unit tested.

File: Logger/BaseLoggerMixins.cs (hunk 1 above)
Comment (connerhanson) [reply Logger/BaseLoggerMixins.cs:21]: I initially meant to do that, but it looks like I forgot to implement it thanks for letting me know!

File: Logger/LogFactory.cs (hunk 2 above)
Comment (connerhanson) [reply Logger/LogFactory.cs]: Thanks for pointing that out should be fixed now.

**Comment by KairyuSen** [comment]
There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔
All of the above should be unit tested.

Looking at your code it looks like you deleted what was initially there in the docs then retyped it, is this the case and if so why? Just curious to the thought process

**Comment by connerhanson** [comment]
@KairyuSen 
As I said above I rm -r'd my initial cloned repo and recloned it to restart. I think the .git file didn't recognize that and shows it as all new code.

File: Logger.Tests/Logger.Tests.csproj [hunk 3]
```diff
@@ -1,9 +1,14 @@
...
+		<TargetFramework>net8.0</TargetFramework>
 		<IsPackable>false</IsPackable>
-		<LangVersion>11.0</LangVersion>
+    <LangVersion>12.0</LangVersion>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <EnableNETAnalyzers>true</EnableNETAnalyzers>
+    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
+    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
+    <Nullable>enable</Nullable>
 	</PropertyGroup>
```
Comment (BenjaminMichaelis) [inline Logger.Tests/Logger.Tests.csproj:12]:
```suggestion
        <LangVersion>12.0</LangVersion>
        <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
        <EnableNETAnalyzers>true</EnableNETAnalyzers>
        <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
        <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
        <Nullable>enable</Nullable>
	</PropertyGroup>
```
spacing should be consistent

File: Logger/BaseLogger.cs [hunk 4]
```diff
@@ -1,7 +1,43 @@
...
-{
-    public abstract void Log(LogLevel logLevel, string message);
-}
-
+﻿using System;
+using System.IO;
+using System.Globalization;
+
+namespace Logger;
+#nullable enable
```
Comment (BenjaminMichaelis) [inline Logger/BaseLogger.cs:6]: Why is these needed?

File: Logger/BaseLogger.cs [hunk 5]
```diff
@@ -1,7 +1,43 @@
...
+
+public abstract class BaseLogger
+{
+    //Auto Property
+    public string? ClassName { get; set; }
```
Comment (BenjaminMichaelis) [inline Logger/BaseLogger.cs:11]: Why is this nullable?

File: Logger/BaseLogger.cs [hunk 6]
```diff
@@ -1,7 +1,43 @@
...
+    {
+        _fileName = fileName;
+    }
+
+
+    //current date/time 
+    private string dateTime = DateTime.Now.ToString("MM/dd/yyyy hh:mm:ss tt", CultureInfo.CurrentCulture);
+
+    //create variable for class name 
+    private string className = nameof(FileLogger);
```
Comment (BenjaminMichaelis) [inline Logger/BaseLogger.cs:32]: Why is this private? We should be using properties instead of fields. Also we should be able to set the class.

File: Logger/BaseLoggerMixins.cs [hunk 7]
```diff
@@ -1,6 +1,45 @@
...
-
-}
+﻿namespace Logger;
+using System.Globalization;
+#nullable enable
+
+public static class BaseLoggerMixins
+{
+
+    public static void Error(this BaseLogger? logger, string message, params object[] arguments)
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:8]: Why is logger nullable? I expect to call this message, a logger is needed.

File: Logger/LogFactory.cs [hunk 8]
```diff
@@ -1,10 +1,34 @@
...
+        //If given null path return null
+        if(path == null)
+        {
+            return null;
+        }
+
+        //If its not null then create new FileLogger and set class name and return the logger
+        else
+        {
+            FileLogger fileLogger = new(ConfigureFileLogger(path));
```
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:29]: Not sure why you need to call ConfigureFileLogger here or what that will do

File: Logger/BaseLogger.cs (hunk 4 above)
Comment (connerhanson) [reply Logger/BaseLogger.cs:6]: I had put these in before enabling nullable in the csproj files.  I'll make sure when refactoring on the next assignment, to take out redundant or unused code before the due date.

File: Logger/LogFactory.cs (hunk 8 above)
Comment (connerhanson) [reply Logger/LogFactory.cs:29]: I put this here to resolve: 

The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ❌ (Didn't use use the configure in the CreateLogger method)

I realize that was not the correct way to implement this step, and that what I had before worked just as well.

File: Logger/BaseLogger.cs (hunk 5 above)
Comment (connerhanson) [reply Logger/BaseLogger.cs:10]: For some reason when removing the nullability the build fails on the pull request I believe this is why I added it. Will investigate more tommorrow.

File: Logger/BaseLoggerMixins.cs (hunk 7 above)
Comment (connerhanson) [reply Logger/BaseLoggerMixins.cs:8]: I did this to resolve: 

"These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started."

I think this is correct since it says if the baseLogger is null throw an exception and you would need to input null for this to happen, though I maybe miss understanding.  I will ask in class tommorrow.
